### PR TITLE
[New Runtime] LibraryManager 2: Electric Boogaloo

### PR DESCRIPTION
**Pull request by anurse**
LibraryManager 2, aka DependencyManager.

Right now it's just a refactoring of the existing code to resolve the dependency graph. Next up though, it will be the place Loaders look for Libraries to load. Just pushing it now so we can review things in nice small chunks

/cc @ChengTian @davidfowl

### CODE REVIEW

**Comment by davidfowl** [comment]
Why are your braces wrong for all of the test projects?

File: src/Microsoft.Framework.ApplicationHost2/Program.cs [hunk 1]
```diff
@@ -114,6 +114,9 @@ private ILoggerFactory ConfigureLogging()
         {
             var loggerFactory = new LoggerFactory();
             loggerFactory.AddConsole(RuntimeLogging.Filter);
+#if DEBUG
+            loggerFactory.MinimumLevel = LogLevel.Debug;
```
Comment (davidfowl) [inline src/Microsoft.Framework.ApplicationHost2/Program.cs]: Does this override what was specified?

File: src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            NuGetFramework targetFramework)
+        {
+            // Walk dependencies
+            var walker = new DependencyWalker(dependencyProviders);
+            var graph = walker.Walk(projectName, version, targetFramework);
+
+            // Resolve conflicts
+            if (!graph.TryResolveConflicts())
+            {
+                throw new InvalidOperationException("Failed to resolve conflicting dependencies!");
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs]: We'll get more specific here in the future

File: src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+            // Build the resolved dependency list
+            var nameMap = new Dictionary<string, Library>();
+            var libraries = new Dictionary<LibraryIdentity, Library>();
+            graph.ForEach(node =>
+            {
+                // Everything should be Accepted or Rejected by now
+                Debug.Assert(node.Disposition != Disposition.Acceptable);
+
+                if(node.Disposition == Disposition.Accepted)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs]: formatting. That vim plugin sucks :stuck_out_tongue_closed_eyes:

File: src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+namespace Microsoft.Framework.Runtime.Dependencies
+{
+    /// <summary>
+    /// Represents the dependency graph for a project
+    /// </summary>
+    public class DependencyManager
+    {
+        private static readonly ILogger Log = RuntimeLogging.Logger<DependencyManager>();
+
+        private readonly GraphNode<Library> _graph;
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs]: What's the purpose of having 3 of these representations around?

**Comment by anurse** [comment]
Wrote 'em in VIM not VS :P. I'll fix them

File: src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs (hunk 4 above)
Comment (anurse) [inline src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs]: They are an artifact of an earlier bit of code I wrote to look things up. I'll remove these for now.

File: src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs (hunk 3 above)
Comment (anurse) [inline src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs]: It's not the vim plugin :)

Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Hosting/Dependencies/DependencyManager.cs]: You're right, Vim is awesome

**Comment by BrennanConroy** [comment]
There are a lot of files missing copyright

**Comment by anurse** [comment]
I'll do a separate pass to copyrightify the code in our repo after merging. I'm not going to do the NuGet ones yet, because I'm not 100% positive on the situation there.

**Comment by muratg** [comment]
@Eilon

**Comment by Eilon** [comment]
Can chat in person on Monday if there are any questions on headers.

**Pull request closed without merging**